src/mcts/search.cc

```diff
@@ -908,22 +971,37 @@ void SearchWorker::ExtendNode(Node* node) {
   // if they are root, then thinking about them is the point.
   if (node != search_->root_node_) {
     if (!board.HasMatingMaterial()) {
-      node->MakeTerminal(GameResult::DRAW);
+      result = GameResult::DRAW;
+      trigger = CertaintyTrigger::TERMINAL;
+      // node->MakeTerminal(GameResult::DRAW);
       return;
     }
 
     if (history_.Last().GetNoCaptureNoPawnPly() >= 100) {
-      node->MakeTerminal(GameResult::DRAW);
+      result = GameResult::DRAW;
+      trigger = CertaintyTrigger::TERMINAL;
+      // node->MakeTerminal(GameResult::DRAW);
       return;
     }
 
     if (history_.Last().GetRepetitions() >= 2) {
-      node->MakeTerminal(GameResult::DRAW);
+      result = GameResult::DRAW;
+      trigger = CertaintyTrigger::TERMINAL;
+      // node->MakeTerminal(GameResult::DRAW);
+      return;
+    }
+
+    if ((history_.Last().GetRepetitions() >= 1) &&
```

Comment:
Twofold repetition is not a draw.
Twofold repetition is useful and people generally like it, we should not remove it.
If it really adds some (>5) Elo, we may consider making it a flag though.

Reply:
I don't know about this. The whole point about CP is to let leela find repetitions earlier. Under a "best play" assumption this is as zero as you can get to address this problem.

 But to clarify: this change does not terminate the game if a two-fold arises, it just scores it as if its a draw. Its not making two-folds terminals, just certain. This means that leela will not waste and dilute visits to otherwise identical positionss

It is responsible for some of the +10 gain of elo, but i really don't know how many. But this is the one thing, that should compound really well in training. I have numerous real world examples from CCCC and TCEC where this finds the correct path to a win, while normal leela stumbles into a perpetual. Sometimes it sees the draw 16 ply earlier (two-fold plus CP) then normal leela. If we assume an effective branching factor of two in chess, we might just need 2^-16 of the training games we otherwise would need for leela to understand these positions. These are all guestimates and no gurantees can be given of course. But it seems like a good try.

The only downside is that you get 0.00 evals if leela sees it can repeat. If the opponent then deviates it will return to the normal eval. Most other engines do this too. But, if you turn CP off, you get normal 3-fold behavior. So I would hate to waste another option on this.